### PR TITLE
Ford Focus Electric support added

### PR DESCRIPTION
**Pull request by martiinezz**
Hi,

I've created FFE related json. Needs testing.

### CODE REVIEW

**Comment by martiinezz** [comment]
Would be nice to get ABRP app with those  new FFE files for testing.

File: ford/FFE.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  "current": {
+    "equation": "((signed(A)*256)+B)*0.02",
+    "minValue": "-500",
+    "maxValue": "300",
+    "type": "Number",
+    "command": "22480B",
+    "ecu": "7E4"
+  },
+  "voltage": {
+    "equation": "((A<<8)+B)*0.01",
```
Comment (SamuelBrucksch) [inline ford/FFE.json]: We also support int16(A:B) now which is a bit nicer to read for this case. We will also update the other json files soon with that.

File: ford/FFE.json [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  "batt_temp": {
+    "equation": "A-50",
+    "minValue": "-40",
+    "maxValue": "80",
+    "type": "Number",
+    "command": "224800",
+    "ecu": "7E4"
+  },
+  "is_charging": {
+    "equation": "A==2",
```
Comment (SamuelBrucksch) [inline ford/FFE.json]:
```suggestion
    "equation": "(A==2)||(A==8)",
```
This should contain both slow and fast charging

File: obdble_cars.json [hunk 3]
```diff
@@ -83,6 +83,11 @@
     "supports": ["calibration", "driving", "charging"],
     "pids": "MachE"
   },
+  "ford:ffe*": {
+    "supports": ["calibration", "driving", "charging"],
+    "pids": "FFE"
+    "groups": ["devel"]
```
Comment (SamuelBrucksch) [inline obdble_cars.json]:
```suggestion
    "groups": ["tester"]
```

**Comment by SamuelBrucksch** [comment]
Thanks for your contribution. I added some comments. Once adjusted we could add it and you could test if it works.

**Comment by SamuelBrucksch** [comment]
Ah one other thing i just noted... Init sequence contains ATSP6 but obd_protocol is 7, this should normally be the same. Which one is correct?

File: ford/FFE.json (hunk 2 above)
Comment (martiinezz) [reply ford/FFE.json]: My 2012 Ford Focus Electric doesn't support DC charging :)

**Comment by martiinezz** [comment]
I based my FFE file on Mach-E - some data is the same (like external temp) - let's try protocol 6 as it's old car.

**Comment by martiinezz** [comment]
I could also provide some more info like odometer, min cell voltage - is it worth doing? I've seen Ioniq5 has much more stuff in json.

**Comment by martiinezz** [comment]
So when will this code show up for testing?

File: ford/FFE.json (hunk 1 above)
Comment (SamuelBrucksch) [inline ford/FFE.json]:
```suggestion
    "equation": "int16(A:B)*0.01",
```

File: ford/FFE.json [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  "is_parked": {
+    "equation": "A==70",
+    "minValue": "0",
+    "maxValue": "1",
+    "type": "Boolean",
+    "command": "221E12",
+    "ecu": "7E2"
+  },
+  "vehicle_reported_speed": {
+    "equation": "((A<<8)+B)/128",
```
Comment (SamuelBrucksch) [inline ford/FFE.json]:
```suggestion
    "equation": "int16(A:B)/128",
```

**Comment by SamuelBrucksch** [comment]
Yes odometer can be used. All of the properties described here can be used:
https://documenter.getpostman.com/view/7396339/SWTK5a8w#fdb20525-51da-4195-8138-54deabe907d5

Please check the telemetry parameters. Let me know if you want to add some, then we can try it out afterwards.

Btw if your car does not fast charge, then we do not need `is_dcfc` as that indicates fast charging. Are there Ford Focus Electric that can fast charge?

**Comment by martiinezz** [comment]
FFE with 23kWh battery had no DC - newer 2017+ had 33,5kWh battery and had CCS. So as ABRP app has 2 types of FFE - then 2012-2016 was without DC but 2017-2018 was equipped with DC.

**Comment by SamuelBrucksch** [comment]
Please send me a mail to samuel@iternio.com with your mail associated to your ABRP account so i can add you the tester role needed to test this. Then we can check if data comes in.

Do you by chance have some sample data that is returned, so i can check before deploying, if everything works? Like in this PR:
https://github.com/iternio/ev-obd-pids/pull/4
This definitely helps to make sure that everything is correct and makes testing easier.

File: ford/FFE.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  "init_commands": {
+    "command": ["ATZ", "ATD", "ATE0", "ATS0", "ATAL", "ATSP6"]
+  },
+  "obd_protocol": "7",
+  "data_commands": {
+    "command": [
+      "ATSH7E2",
+      "224850",
+      "224851",
+      "224845"
```
Comment (SamuelBrucksch) [inline ford/FFE.json]:
```suggestion
      "224845",
```

File: ford/README.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
 # Ford Mustang Mach E OBD Data
 
 These OBD PIDs are based on the work by Ben Karleskint
+
+# Ford Focus Electric 2012
+
+Based on https://revisto.eu/
```
Comment (SamuelBrucksch) [inline ford/README.md:7]: Maybe add your name here? That website does not contain much information...

Comment (martiinezz) [reply ford/README.md:7]: I'll update website soon (next week)

**Comment by martiinezz** [comment]
I made an error as for voltage we need to replace 224850 with 22480D

File: ford/FFE.json [hunk 7]
```diff
@@ -0,0 +1,92 @@
+{
+  "init_commands": {
+    "command": ["ATZ", "ATD", "ATE0", "ATS0", "ATAL", "ATSP6"]
+  },
+  "obd_protocol": "7",
+  "data_commands": {
+    "command": [
+      "ATSH7E2",
+      "224850",
```
Comment (martiinezz) [inline ford/FFE.json:9]: should be 22480D

File: ford/FFE.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    "type": "Number",
+    "command": "22480B",
+    "ecu": "7E4"
+  },
+  "voltage": {
+    "equation": "int16(A:B)*0.01",
+    "minValue": "200",
+    "maxValue": "400",
+    "type": "Number",
+    "command": "224850",
```
Comment (martiinezz) [inline ford/FFE.json:33]: should be 22480D

**Pull request closed without merging**